estimagic/inference/shared.py

```diff
@@ -0,0 +1,118 @@
+            see :ref:`implementation_of_constraints`.
+        constraints (list): List with constraint dictionaries.
+            See .. _link: ../../docs/source/how_to_guides/how_to_use_constraints.ipynb
+        n_samples (int): Number of samples used to transform the covariance matrix of
+            the internal parameter vector into the covariance matrix of the external
+            parameters.
+        bounds_handling (str): One of "clip", "raise", "ignore". Determines how bounds
+            are handled. If "clip", confidence intervals are clipped at the bounds.
+            Standard errors are only adjusted if a sampling step is necessary due to
+            additional constraints. If "raise" and any lower or upper bound is binding,
+            we raise an error. If "ignore", boundary problems are simply ignored.
+
+    Returns:
+        pd.DataFrame: Quadratic DataFrame containing the covariance matrix of the free
+            parameters. If parameters were fixed (explicitly or by other constraints),
+            the index is a subset of params.index. The columns are the same as the
+            index.
+
+    """
+    processed_constraints, processed_params = process_constraints(constraints, params)
+    free_index = processed_params.query("_internal_free").index
+
+    if processed_constraints:
+        free = processed_params.loc[free_index]
+        is_free = processed_params["_internal_free"].to_numpy()
+        pre_replacements = processed_params["_pre_replacements"].to_numpy()
+        post_replacements = processed_params["_post_replacements"].to_numpy()
+        fixed_values = processed_params["_internal_fixed_value"].to_numpy()
+        lower_bounds = free["_internal_lower"]
+        upper_bounds = free["_internal_upper"]
+
+        internal_mean = reparametrize_to_internal(
+            external=params["value"].to_numpy(),
+            internal_free=is_free,
+            processed_constraints=processed_constraints,
+        )
+        sample = np.random.multivariate_normal(
+            mean=internal_mean, cov=internal_cov, size=n_samples,
+        )
+        transformed_free = []
+        for params_vec in sample:
+            if bounds_handling == "clip":
+                params_vec = np.clip(params_vec, a_min=lower_bounds, a_max=upper_bounds)
+            elif bounds_handling == "raise":
+                if (params_vec < lower_bounds).any() or (
+                    params_vec > upper_bounds
+                ).any():
+                    raise ValueError()
+
+            transformed = reparametrize_from_internal(
+                internal=params_vec,
+                fixed_values=fixed_values,
+                pre_replacements=pre_replacements,
+                processed_constraints=processed_constraints,
+                post_replacements=post_replacements,
+            )
+            transformed_free.append(transformed[is_free])
+
+            free_cov = np.cov(np.array(transformed_free), rowvar=False,)
```

Comment:
I think you can indent `free_cov = np.cov(np.array(transformed_free), rowvar=False,)` one level out and just calculate the covariance once after the loop.

Reply:
Done